core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala

```diff
@@ -0,0 +1,251 @@
+     * is kept and updated upon each invocation.
+     *
+     * If asked, check for sentinel markers - but exclude the identified markers from
+     * the result returned from this method.
+     *
+     * Only parses and returns as much logs as fit in the passed log limit.
+     * Even if the log limit is exceeded, advance the starting position for the next invocation
+     * behind the bytes most recently read - but don't actively read any more until sentinel
+     * markers have been found.
+     *
+     * @param limit the limit to apply to the log size
+     * @param waitForSentinel determines if the processor should wait for a sentinel to appear
+     *
+     * @return a vector of Strings with log lines in our own JSON format
+     */
+    def logs(limit: ByteSize, waitForSentinel: Boolean)(implicit transid: TransactionId): Future[Vector[String]] = {
+
+        def readLogs(retries: Int): Future[Vector[String]] = {
+            docker.rawContainerLogs(id, logFileOffset).flatMap { rawLogBytes =>
+                val rawLog = new String(rawLogBytes.array, rawLogBytes.arrayOffset, rawLogBytes.position, StandardCharsets.UTF_8)
+                val (isComplete, isTruncated, formattedLogs) = processJsonDriverLogContents(rawLog, waitForSentinel, limit)
+
+                if (retries > 0 && !isComplete && !isTruncated) {
+                    log.info(this, s"log cursor advanced but missing sentinel, trying $retries more times")
+                    Thread.sleep(logsRetryWait.toMillis)
+                    readLogs(retries - 1)
+                } else {
+                    logFileOffset += rawLogBytes.position - rawLogBytes.arrayOffset
+                    Future.successful(formattedLogs)
+                }
+            }.andThen {
+                case Failure(e) =>
+                    log.error(this, s"Failed to obtain logs of ${id.asString}: ${e.getClass} - ${e.getMessage}")
+            }
+        }
+
+        readLogs(logsRetryCount)
+    }
+
+    /**
+     * Makes an HTTP request to the container.
+     *
+     * Note that `http.post` will not throw an exception, hence the generated Future cannot fail.
+     *
+     * @param path relative path to use in the http request
+     * @param body body to send
+     * @param timeout timeout of the request
+     * @param retry whether or not to retry the request
+     */
+    protected def callContainer(path: String, body: JsObject, timeout: FiniteDuration, retry: Boolean = false): Future[RunResult] = {
+        val started = Instant.now()
+        val http = httpConnection.getOrElse {
+            val conn = new HttpUtils(s"${ip.asString}:8080", timeout, 1.MB)
+            httpConnection = Some(conn)
+            conn
+        }
+        Future {
+            http.post(path, body, retry)
+        }.map { response =>
+            val finished = Instant.now()
+            RunResult(Interval(started, finished), response)
+        }
```

Comment:
Besides the access scope change, we left `HttpUtils` as is. It still returns a `Left()` if unsuccessful or a `Right()` if successful.

Both callers of the code under discussion here - namely, `initialize()` and `run()` - have the following error handling code:

```
        callContainer("/init", body, timeout, retry = true).andThen { // never fails
            case Success(r: RunResult) =>
                transid.finished(this, start.copy(start = r.interval.start), s"initialization result: ${r.toBriefString}", endTime = r.interval.end)
            case Failure(t) =>
                transid.failed(this, start, s"initializiation failed with $t")
        } ...
```

So we are going to log a failed `Future` returned by `callContainer()` and return the failed `Future` to the caller of `initialize()` and `run()`. Looking at https://github.com/openwhisk/openwhisk/pull/2021/files, I do NOT think that the caller of `initialize()` and `run()` fully addresses a failed `Future`.

@rabbah do you suggest to `recover` here at line 251 and transform the `Failure(t)` into a `RunResult(..., Left(t))`? This transformation would bypass the code snippet from above with `transid.failed()`. I guess we need a decision how hard the penalty for a failed `Future` is: a log statement, a failed transaction marker, a failed init / run, or even an assertion stopping the invoker process. What do you have in mind?